src/main/java/com/exactpro/th2/crawler/Crawler.java

```diff
@@ -383,6 +468,64 @@ private SendingReport sendMessages(CrawlerId crawlerId, DataServiceInfo dataServ
         return new SendingReport(CrawlerAction.NONE, dataServiceName, dataServiceVersion);
     }
 
+    private Map<String, RecoveryState.InnerMessage> mapAliasesWithMessages(Collection<MessageData> messages) {
+        RecoveryState.InnerMessage message;
+        Map<String, RecoveryState.InnerMessage> result = new HashMap<>();
+
+        for (MessageData messageData : messages) {
+
+            String alias = messageData.getMessageId().getConnectionId().getSessionAlias();
+            Instant timestamp = Instant.ofEpochSecond(messageData.getTimestamp().getSeconds(), messageData.getTimestamp().getNanos());
+            Direction direction = Direction.valueOf(messageData.getDirection().toString());
+            long sequence = messageData.getMessageId().getSequence();
+
+            message = new RecoveryState.InnerMessage(alias, timestamp, direction, sequence);
+
+            result.put(alias, message);
+        }
+
+        return result;
+    }
+
+    private Map<String, MessageID> getLastMessagesWithAliases(List<MessageData> messages) {
+        Map<String, MessageID> result = new HashMap<>();
+
+        Set<String> aliases = messages.stream().filter(MessageData::hasMessageId)
+                .map(MessageData::getMessageId)
+                .map(MessageID::getConnectionId)
+                .map(ConnectionID::getSessionAlias)
+                .collect(Collectors.toSet());
```

Comment:
You can group messages right here.
`Collectors.toMap` has overloads with lambda that allows you to merge the previous and a new value the way you want